### PR TITLE
Attempt #3: Adding the ability to upload tours to Strava

### PR DESCRIPTION
**Pull request by FJBDev**
Per the [Feature Request](https://sourceforge.net/p/mytourbook/feature-requests/150/), MyTourbook is now integrated with [Strava](https://www.strava.com) and allows uploading Tours on your Strava account.

If you don't have a Strava account, you can create one for free [here](https://www.strava.com/register/free)

### CONNECT TO YOUR STRAVA ACCOUNT
- Go to Preferences :arrow_right: Cloud Connectivity :arrow_right: Strava
- Click on the button "Connect to Strava"

- Enter your Strava credentials

- Click on "Authorize"

- A dialog should display that the authorization was successful. If not, it will display the error message returned by Strava.

- The preferences will show information about the Strava account that will be used to upload tours as well as technical information such as access token, refresh token...etc

### UPLOAD TOURS TO YOUR STRAVA ACCOUNT

- In the Tour Book View, select 1 or more tours and right click on **"Upload Tours" :arrow_right: "Strava"**

- A dialog will show the progress of the upload to Strava.

image

- The summary will list the uploaded tours as well as the tours that could not be uploaded. To know the details of the reasons why some tours were not successfully uploaded, consult the Log View.

image

- @wolfgang-ch : I have added unit tests for both [GPX](https://github.com/FJBDev/mytourbook/blob/c4f458b3e1f6747696ca42f29f11bce9cf997e67/bundles/net.tourbook.tests/test/exportdata/gpx/ExportGpxTester.java#L37) and [TCX](https://github.com/FJBDev/mytourbook/blob/c4f458b3e1f6747696ca42f29f11bce9cf997e67/bundles/net.tourbook.tests/test/exportdata/garmin/tcx/ExportTcxTester.java#L37). Another case where I think unit that tests are very important (if not mandatory) as they helped ensuring that I didn't break the tour export GPX/TCX 😃 
Also, I have squashed my work in a single commit as you can see here : [7dfdec0](https://github.com/wolfgang-ch/mytourbook/pull/273/commits/7dfdec05b887b34b54104cddce8f765b245563ab)

@wolfgang-ch In the Strava [guidelines](https://developers.strava.com/guidelines/), I read this :
_"All apps must display the “Powered by Strava” logo or “Compatible with Strava” logo on all websites, apps and any distributable media such as images, videos or prints. No variations or modifications are acceptable."_

I am thinking you could display the logo "Compatible With Strava" on the website in the next release and that would comply with their requirements ?
image

### CODE REVIEW

**Review by wolfgang-ch — COMMENTED** [review]
There was another exception but I could not repeat it. It was initiated from the upload action, which can easily be fixed with an isDispose check.

```
org.eclipse.swt.SWTException: Widget is disposed
	at org.eclipse.swt.SWT.error(SWT.java:4869)
	at org.eclipse.swt.SWT.error(SWT.java:4784)
	at org.eclipse.swt.SWT.error(SWT.java:4755)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:432)
	at org.eclipse.swt.widgets.Widget.getDisplay(Widget.java:554)
	at net.tourbook.ui.views.tourBook.TourBookView.getSelectedTours(TourBookView.java:2040)
	at net.tourbook.extension.upload.ActionUpload$ActionUploadTour.run(ActionUpload.java:77)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:579)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:413)
	at org.eclipse.jface.action.ActionContributionItem$$Lambda$331/000000000000000000.handleEvent(Unknown Source)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4195)
```

Finally the upload worked :-)

File: bundles/net.tourbook.cloud/src/net/tourbook/cloud/oauth2/OAuth2BrowserDialog.java [hunk 1]
```diff
@@ -96,13 +105,12 @@ protected Control createDialogArea(final Composite parent) {
             .applyTo(displayArea);
 
       final Browser browser = new Browser(displayArea, SWT.NONE);
-      GridDataFactory.fillDefaults().grab(true, true).applyTo(browser);
-      browser.setUrl(url);
+      browser.setUrl(_url);
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook.cloud/src/net/tourbook/cloud/oauth2/OAuth2BrowserDialog.java:108]: Oh good catch. I may have done that when I was playing with it to make Dropbox work

File: bundles/net.tourbook/src/net/tourbook/Messages.java [hunk 2]
```diff
@@ -445,6 +445,7 @@
    public static String        action_tourbook_select_year_month_tours;
    public static String        action_tourCatalog_open_compare_wizard;
    public static String        action_tourType_modify_tourTypes;
+   public static String        action_upload_tour;
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/Messages.java:448]: Fixed in next PR

File: bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/StravaUploader.java [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+public class StravaUploader extends TourbookCloudUploader {
+
+   private static HttpClient       httpClient     = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(20)).build();
+   private static final String     _stravaBaseUrl = "https://www.strava.com/api/v3/";                                      //$NON-NLS-1$
+
+   public static final String      HerokuAppUrl   = "https://passeur-mytourbook-strava.herokuapp.com";                     //$NON-NLS-1$
+
+   private static IPreferenceStore _prefStore     = Activator.getDefault().getPreferenceStore();
+   private static TourExporter     _tourExporter  = new TourExporter(ExportTourTCX.TCX_2_0_TEMPLATE);
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/StravaUploader.java:79]: Fixed in next PR

File: bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/PrefPageStrava.java [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+   private Label  _labelAthleteName;
+   private Label  _athleteFullName;
+   private Label  _labelAthleteWebPage;
+   private Link   _athleteWebPageLink;
+   private Label  _labelAccessToken;
+   private Label  _accessToken;
+   private Label  _labelRefreshToken;
+   private Label  _refreshToken;
+   private Label  _labelExpiresAt;
+   private Label  _labelAccessTokenExpiresAt;
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/PrefPageStrava.java:68]: They already have a common structure : prefix of "_label" for the static labels and no prefix for the labels containing the data.
Feel free to suggest a better naming structure

File: bundles/net.tourbook/src/net/tourbook/ui/views/tourBook/TourBookView.java [hunk 5]
```diff
@@ -217,11 +218,11 @@
 
 // SET_FORMATTING_ON
    //
-   static public final String ID = "net.tourbook.views.tourListView"; //$NON-NLS-1$
+   public static final String ID = "net.tourbook.views.tourListView"; //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/ui/views/tourBook/TourBookView.java:221]: Fixed in next PR (see [SO](https://stackoverflow.com/questions/5313470/getting-velocity-error-at-statup-vm-global-library-vm))

**Comment by FJBDev** [comment]
> There was another exception but I could not repeat it. It was initiated from the upload action, which can easily be fixed with an isDispose check.
> 
> ```
> org.eclipse.swt.SWTException: Widget is disposed
> 	at org.eclipse.swt.SWT.error(SWT.java:4869)
> 	at org.eclipse.swt.SWT.error(SWT.java:4784)
> 	at org.eclipse.swt.SWT.error(SWT.java:4755)
> 	at org.eclipse.swt.widgets.Widget.error(Widget.java:432)
> 	at org.eclipse.swt.widgets.Widget.getDisplay(Widget.java:554)
> 	at net.tourbook.ui.views.tourBook.TourBookView.getSelectedTours(TourBookView.java:2040)
> 	at net.tourbook.extension.upload.ActionUpload$ActionUploadTour.run(ActionUpload.java:77)
> 	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
> 	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:579)
> 	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:413)
> 	at org.eclipse.jface.action.ActionContributionItem$$Lambda$331/000000000000000000.handleEvent(Unknown Source)
> 	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
> 	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4195)
> ```
> 
> Finally the upload worked :-)

I can't reproduce the error...
Looking at the stack trace, it seems that the issue is in the TourBookView but not in the Strava feature code

File: bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/PrefPageStrava.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wolfgang-ch) [reply bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/PrefPageStrava.java:68]: These are my field names suggestions...

mt-fields-2


...further I would put fields, which belong together, into { ... }, for me it is better readable and has the advantage, that you can select fields more easily with `<Alt><Shift> + <Left> or <Right>` when you rearrange them, which I did very often especially in the tour editor

mt-code-layout

**Comment by wolfgang-ch** [comment]
> it seems that the issue is in the TourBookView but not in the Strava feature code

As I pointed out above, it was initiated from the upload action

`at net.tourbook.extension.upload.ActionUpload$ActionUploadTour.run(ActionUpload.java:77)`

but I don't understand why it occurred because I don't see a delayed action which could cause this exception, I fixed it in https://github.com/wolfgang-ch/mytourbook/commit/e4ce6b230c10a1bdeb654f98534395b3896e6d1f

File: bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/PrefPageStrava.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (FJBDev) [reply bundles/net.tourbook.cloud/src/net/tourbook/cloud/strava/PrefPageStrava.java:68]: Thanks for the suggestions. I implemented them in my next PR.

**Pull request merged**